src/lib/components/InfiniteScroll.svelte

```diff
@@ -0,0 +1,170 @@
+<script lang="ts">
+    import { onMount } from 'svelte';
+    import { PAGE_SIZE } from '$lib';
+    import { toast } from 'svelte-sonner';
+    import type { PostData, uuid } from '$lib/types';
+    import { viewport } from '$lib/util';
+
+    import Clock from 'lucide-svelte/icons/clock';
+    import ArrowUp from 'lucide-svelte/icons/arrow-up';
+    import ChevronUp from 'lucide-svelte/icons/chevron-up';
+    import ChevronDown from 'lucide-svelte/icons/chevron-down';
+
+    import Post from './Post.svelte';
+
+    import * as Select from '$lib/shadcn/components/ui/select';
+    import { Button } from '$lib/shadcn/components/ui/button';
+
+    export let data: { signed_in?: boolean } & (
+        | { source: 'home' }
+        | { source: 'channel'; channel_name: string }
+        | { source: 'private_channel'; channel_id: uuid }
+        | { source: 'user'; username: string }
+    );
+
+    const url =
+        data.source === 'home'
+            ? '/api/posts'
+            : data.source === 'channel'
+              ? `/api/c/${data.channel_name}/posts`
+              : data.source === 'private_channel'
+                ? `/api/c/private/${data.channel_id}/posts`
+                : data.source === 'user'
+                  ? `/api/u/${data.username}/posts`
+                  : 0; // unreachable
+
+    let filter = {
+        value: 'all',
+        label: 'All',
+        disabled: false,
+    };
+
+    let sort = {
+        value: 'votes',
+        label: 'Top',
+        disabled: false,
+    };
+
+    let reverseSort = false;
+
+    let page: number = 0;
+    let buffer: PostData[] = [];
```

Comment:
buffer should be exposed to parent and clear-able from the parent component (refresh button in the future)